design/general-progress-monitoring.md

```diff
@@ -0,0 +1,462 @@
+    }
+
+Two new methods will be added to the RestoreItemAction interface, and the
+RestoreItemActionExecuteOutput struct will be modified:
+
+	// Execute allows the ItemAction to perform arbitrary logic with the item being restored,
+	// including mutating the item itself prior to restore. The item (unmodified or modified)
+	// should be returned, an optional OperationID, along with an optional slice of ResourceIdentifiers
+	// specifying additional related items that should be restored, a warning (which will be
+	// logged but will not prevent the item from being restored) or error (which will be logged
+	// and will prevent the item from being restored) if applicable. If OperationID is specified
+    	// then velero will wait for this operation to complete before the restore is marked Completed.
+	Execute(input *RestoreItemActionExecuteInput) (*RestoreItemActionExecuteOutput, error)
+
+    	
+    	// Progress  
+    	Progress(input *RestoreItemActionProgressInput) (OperationProgress, error)
+
+        // Cancel  
+    	Cancel(input *RestoreItemActionProgressInput) error
+    
+    // RestoreItemActionExecuteOutput contains the output variables for the ItemAction's Execution function.
+    type RestoreItemActionExecuteOutput struct {
+        // UpdatedItem is the item being restored mutated by ItemAction.
+        UpdatedItem runtime.Unstructured
+
+        // AdditionalItems is a list of additional related items that should
+        // be restored.
+        AdditionalItems []ResourceIdentifier
+
+        // SkipRestore tells velero to stop executing further actions
+        // on this item, and skip the restore step. When this field's
+        // value is true, AdditionalItems will be ignored.
+        SkipRestore bool
+
+        // OperationID is an identifier which indicates an ongoing asynchronous action which Velero will
+        // continue to monitor after restoring this item. If left blank, then there is no ongoing operation
+        OperationID string
+    }
+
+    // RestoreItemProgressInput contains the input parameters for the RestoreItemAction's Progress function.
+    type RestoreItemProgressInput struct {
+    	// Item is the item that was stored in the restore
+    	Item runtime.Unstructured
+    	// OperationID is the operation ID returned by RestoreItemAction Execute
+    	operationID string
+    	// Restore is the representation of the restore resource processed by Velero.
+    	Restore *velerov1api.Restore
+    }
+
+## Changes in Velero backup format
+
+No changes to the existing format are introduced by this change.  A
+`<backup-name>-itemoperations.json.gz` file will be added that contains the items and operation IDs
+(snapshotIDs) returned by VolumeSnapshotter and BackupItemAction plugins.  Also, the creation of the
+`velero-backup.json` object will not occur until the backup moves to one of the terminal phases
+(_Completed_, _PartiallyFailed_, or _Failed_).  Reconciliation should ignore backups that do not
```

Comment:
As previous comments, I prefer to filter backups by Phase, not the object file

Reply:
Looking at this again, I'm not actually 100% sure what @dsu-igeek meant here in the original design. Once we have that file, (unless we change things in response to the earlier suggestion  to write it sooner), the backup is already completed, so I'm not sure why this means that the controller reconcile would ignore backups without the file -- in fact, the opposite might be true.